File: src/org/elixir_lang/credo/Configurable.java
package org.elixir_lang.credo;

import com.intellij.execution.configuration.EnvironmentVariablesComponent;
import com.intellij.openapi.options.SearchableConfigurable;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class Configurable implements SearchableConfigurable {
    private static final String ID = "language.elixir.credo";
    private static final String DISPLAY_NAME = "Credo";
    private JPanel panel;
    private JTextField elixirArguments;
    private JTextField erlArguments;
    private EnvironmentVariablesComponent environmentVariablesComponent;
    private Service service;

    public Configurable(@NotNull Project project) {
        service = Service.getInstance(project);
    }

    @NotNull
    @Override
    public String getDisplayName() {
        return DISPLAY_NAME;
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        return panel;
    }

    @Override
    public boolean isModified() {
        return !service.getElixirArguments().equals(elixirArguments.getText()) ||
                !service.getErlArguments().equals(erlArguments.getText()) ||
                !service.getEnvironmentVariableData().equals(environmentVariablesComponent.getEnvData());
    }

    @Override
    public void apply() {
        service.setElixirArguments(elixirArguments.getText());
        service.setErlArguments(erlArguments.getText());
        service.setEnvironmentVariableData(environmentVariablesComponent.getEnvData());
    }

    @Override
    public void reset() {
        elixirArguments.setText(service.getElixirArguments());
        erlArguments.setText(service.getErlArguments());
        environmentVariablesComponent.setEnvData(service.getEnvironmentVariableData());
    }

    @NotNull
    @Override
    public String getId() {
        return ID;
    }
}

File: src/org/elixir_lang/dialyzer/service/Configurable.kt
package org.elixir_lang.dialyzer.service

import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.project.Project
import org.jetbrains.annotations.Nls
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.JTextField

class Configurable(project: Project) : SearchableConfigurable {
    private val service: DialyzerService = project.getService(DialyzerService::class.java)
    private var mixArguments: JTextField? = null
    private var elixirArguments: JTextField? = null
    private var erlArguments: JTextField? = null
    private var panel: JPanel? = null
    override fun getId() = "Dialyzer"

    @Nls(capitalization = Nls.Capitalization.Title)
    override fun getDisplayName(): String = id

    override fun createComponent(): JComponent? = panel

    override fun isModified(): Boolean =
            (service.mixArguments != mixArguments?.text) ||
            (service.elixirArguments != elixirArguments?.text) ||
            (service.erlArguments != erlArguments?.text)

    override fun apply() {
        mixArguments?.text?.let {
            service.mixArguments = it
        }
        elixirArguments?.text?.let {
            service.elixirArguments = it
        }
        erlArguments?.text?.let {
            service.erlArguments = it
        }
    }

    override fun reset() {
        mixArguments?.text = service.mixArguments
        elixirArguments?.text = service.elixirArguments
        erlArguments?.text = service.erlArguments
    }
}
